WordSearch.java

```diff
@@ -0,0 +1,42 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class WordSearch {
+    private List<String> results = new ArrayList<>();

```

Comment:
It would be better to pass results as a parameter to the methods rather than have it as a member of class.

What will happen if someone decides to call findAllWords on the same WordSearch object twice with different grids (there is no comment that would forbid it)?

Usually it is better to keep classes stateless (as in: after an object is constructed, no matter how many times methods are called and in what order, object will not change and each invocation of a method will yield the same result) unless you have very good reason to do otherwise (like: caching or sth), since it lets you avoid really nasty bugs that appear only in very specific circumstances (specific sequence of events) and are very hard to debug.

On the other hand `Dictionary` could be safely assumed to be immutable and thus be assigned to a final field in constructor ("this instance of `WordSearch` works only with this dictionary") and not be passed around in all calls.